File: working/server/php/public_html/att/callback.php
<?php

require_once("../config.php");

#
# The oAuth Consent Request page hands off to this routine after the user has either entered their credentials or
# has denied the application access to the requested functionality.
#
# If authorized, a request code is passed in the request body.
# If denied, an error is generated and submitted along with an error description.
#

if (isset($_GET['code'])) {

    # Once the user has logged-in with their credentials, they get re-directed to this URL
    # with a 'code' parameter. This is exchanged for an access token which can be used in any
    # future calls to the AT&T APIs

    $code = trim($_GET['code']);

    # First, make sure to send the correct content-type
    header('Content-Type: text/html');

    if (!$code) {
        echo  REDIRECT_HTML_PRE . '{"success": false,"msg": "No auth code "}'. REDIRECT_HTML_POST;
    } else {
        $response = $provider->getToken($code);

        if ($response->isError()) {
            echo  REDIRECT_HTML_PRE . '{"success": false,"msg": "Could not fetch auth token"}'. REDIRECT_HTML_POST;
        } else {
            # Store the auth token in the session for use in future API calls

            if (isset($_GET['scopes'])) {
                $scopes = explode(",", $_GET['scopes']);
                foreach ($scopes as $key => $value) {
                    $_SESSION['tokens'][$value] = $response->data()->access_token;
                }
            }
            $_SESSION['refresh_token'] = $response->data()->refresh_token;

            echo  REDIRECT_HTML_PRE . '{"success": true,"msg": "Process Callback"}' . REDIRECT_HTML_POST;
        }
    }
}
else {
    $error = $_REQUEST['error'];
    $error_reason = $_REQUEST['error_reason'];
    $error_description = $_REQUEST['error_description'];

    echo REDIRECT_HTML_PRE . '{"success" : false, "msg" : "$error_description"}' . REDIRECT_HTML_POST;
}

?>

File: working/server/ruby/example/callback.rb
# Once the user has logged in with their credentials, they get redirected to this URL
# with a 'code' and a 'scopes' parameters. This is exchanged for an access token which can be used in any
# future calls to the AT&T APIs.
get '/att/callback' do

  puts "handle callback"

  session[:NAME]  = "CONCHADETURAOSIDH"
  if !params[:code]
    puts 'callback : No auth code on querystring'

    content_type 'text/html'
    response = {
      :success => false,
      :msg => "No auth code"
    }
    
    if params[:error]
        response.merge!({
          :msg => params[:error_description],
          :code => params[:error]
        })
    end

  else

    puts "getToken"
    response = @@att.getToken(params[:code])
  
    if response.error?
      puts "callback : error in response"
  
      content_type 'text/html'
      response = {
        :success => false,
        :msg => "Process Callback",
        :error => response.error
      }
  
    else
      puts "callback : response good"
      # This stores the OAuth token in the session for use in future API calls.
      tokenMap = session[:tokenMap] || {} 
     
      scopes = params[:scopes]

      scopes.split(",").each {|scope|
        tokenMap[scope] = response.data['access_token']
        puts "saved #{scope}: #{tokenMap[scope]}"
      } unless scopes.nil?
      
      session['tokenMap'] = tokenMap

      session['refresh_token'] = response.data['refresh_token']
      puts "refresh_token = #{session[:refresh_token]}"
  
      content_type 'text/html'
      response = {
        :success => true,
        :msg => "Process Callback"
      }
    end
    
  end
  # This sends back our json wrapped in html.
  return REDIRECT_HTML_PRE + response.to_json + REDIRECT_HTML_POST

end
